lib/services/task.ts

```diff
@@ -247,4 +278,79 @@ export class TaskService {
       return { success: false, error: 'Failed to update task order' };
     }
   }
+  // Sync tasks
+  // Pulls pending changes from Dexie and sends them to the server
+  // Marks Dexie tasks synced after successful sync
+  // returns the count of synced tasks
+  async syncTasks(userId: string): Promise<ServiceResponse<number>> {
+    try {
+      // fetch:
+      // New tasks - sync status new - separate new and pending? Use the same one?
+      const newTasks = await db.tasks
+        .where('syncStatus')
+        .equals('new')
+        .toArray();
+      // Updated tasks - sync status pending
+      const updatedTasks = await db.tasks
+        .where('syncStatus')
+        .equals('pending')
+        .toArray();
+      // Deleted tasks - sync status deleted
+      const deletedTasks = await db.tasks
+        .where('syncStatus')
+        .equals('deleted')
+        .toArray();
+
+      // Handle Task Tags
+      const newTaskTags = await db.taskTags
+        .where('taskId')
+        .anyOf(newTasks.map((t) => t.id))
+        .toArray();
+      const updatedTaskTags = await db.taskTags
+        .where('taskId')
+        .anyOf(updatedTasks.map((t) => t.id))
+        .toArray();
+
+      // Hit 'api/tasks/' endpoint, POST with the above data
+      const response = await fetch('/api/tasks', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({
+          userId,
+          newTasks,
+          updatedTasks,
+          newTaskTags,
+          updatedTaskTags,
+          deletedTasks,
+        }),
+      });
+      // if response not okay - throw error
+      if (!response.ok) throw new Error('Failed to sync tasks');
+
+      // Delete removed tasks from Dexie
+      await db.tasks.where('syncStatus').equals('deleted').delete();
+      // Mark synced tasks in Dexie
+      // Something like: db.tasks.where('syncStatus').anyof([*statuses*]).modify({syncStatus: 'synced'})
+      const updated = await db.tasks
+        .where('syncStatus')
+        .anyOf(['new', 'pending'])
+        .modify({ syncStatus: 'synced' });
+
+      return { success: true, data: updated };
+    } catch (error) {
+      return { success: false, error: `Failed to sync tasks: ${error}` };
+    }
+  }
 }
+
+/** Our Task Scenarios:
+ *
+ * 1. A task is added, it receives 'new' sync status
+ *   1 a. Is a 'new' task is updated - it remains 'new' - it will not exist in the AWS db yet
+ *   2 b. Is a 'new' task is deleted - it's completely removed, it will not exist in the AWS db, and does not need to be synced
+ * 2. A task synced with AWS is updated, it receives 'pending' sync status
+ *  2 a. Is a 'pending' task is deleted - it's marked deleted
+ * 3. A task is deleted, it receives 'deleted' sync status - to support this we need to rethink how we handle deleting tasks above
+ */
```

Comment:
This will probably grow, and we should think on a nice functional approach to this, to hopefully keep the method easy to read. 

Like what happens when we just update the tags?